tests/future/batch_semaphore.rs

```diff
@@ -0,0 +1,654 @@
+                    // after the acquire, we are causally dependent on both tasks
+                    check_clock(|i, c| (i == 3) || (c > 0));
+                });
+            },
+            None,
+        );
+    }
+}
+
+#[test]
+fn batch_semaphore_clock_3() {
+    for fairness in [Fairness::StrictlyFair, Fairness::Unfair] {
+        check_dfs(
+            move || {
+                let s = Arc::new(BatchSemaphore::new(0, fairness));
+
+                for i in 1..=2 {
+                    let s2 = s.clone();
+                    thread::spawn(move || {
+                        assert_eq!(me(), i);
+                        s2.release(1);
+                    });
+                }
+
+                thread::spawn(move || {
+                    assert_eq!(me(), 3);
+                    check_clock(|i, c| (c > 0) == (i == 0));
+                    // acquire 1: unblocked once either of the threads finished
+                    s.acquire_blocking(1).unwrap();
+                    // after the acquire, we are causally dependent on exactly one of the two tasks
+                    let clock = current::clock();
+                    assert!((clock[1] > 0 && clock[2] == 0) || (clock[1] == 0 && clock[2] > 0));
+                });
+            },
+            None,
+        );
+    }
+}
+
+#[test]
+fn batch_semaphore_clock_4() {
+    for fairness in [Fairness::StrictlyFair, Fairness::Unfair] {
+        check_dfs(
+            move || {
+                let s = Arc::new(BatchSemaphore::new(1, fairness));
+
+                for tid in 1..=2 {
+                    let other_tid = 2 - tid;
+                    let s2 = s.clone();
+                    thread::spawn(move || {
+                        assert_eq!(me(), tid);
+                        match s2.try_acquire(1) {
+                            Ok(()) => {
+                                // we won the race, no causal dependence on another thread
+                                check_clock(|i, c| (c > 0) == (i == 0 || i == tid));
+                            }
+                            Err(TryAcquireError::NoPermits) => {
+                                // we lost the race, so we causally depend on the other thread
+                                check_clock(|i, c| !(i == 0 || i == other_tid) || (c > 0));
+                            }
+                            Err(TryAcquireError::Closed) => unreachable!(),
+                        }
+                    });
+                }
+            },
+            None,
+        );
+    }
+}
+
+/// Shows a case in which causality tracking in the batch semaphore is
+/// currently imprecise. The test sets up a semaphore with two permits and two
+/// threads, each of which acquires one permit, then releases one permit, then
+/// repeats. Neither thread can be blocked in this situation, and so neither
+/// thread should causally depend on the other, but currently they do.
+#[test]
+#[should_panic(expected = "doesn't satisfy predicate")]
```

Comment:
This shouldn't be `should_panic`, but `ignore` no ?

Reply:
Yes. Why do we want to waste CPU cycles to confirm that we have not improved Shuttle? I like the mechanical-ness of "oh hey btw here's a test which now works", but that could also be gotten by running this test (and optionally other tests) in CI.

Not blocking for me, I'll leave it for others to decide.

Reply:
Actually, I asked Aurel to add this, so we have a scenario showing that our causality tracking is imprecise.

Reply:
> Actually, I asked Aurel to add this, so we have a scenario showing that our causality tracking is imprecise.

?

I don't disagree on the test but whether it should be `ignore` (and not `should_panic`) or `should_panic` (and not `ignore`)

Reply:
My two cents: with the explanation in the comment, the risk of someone mistaking this test as desirable behavior is pretty low. I think it's better to run the test than not to run it. If the test starts to break without addressing the issue, we'd like to know about it, instead of hoping that it still does what we want when running it manually.